liota/dccs/iotcv2.py

```diff
@@ -0,0 +1,346 @@
+#                                                                             #
+#  - Redistributions in binary form must reproduce the above copyright        #
+#      notice, this list of conditions and the following disclaimer in the    #
+#      documentation and/or other materials provided with the distribution.   #
+#                                                                             #
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"#
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE  #
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE #
+#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE  #
+#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR        #
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF       #
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS   #
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN    #
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)    #
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF     #
+#  THE POSSIBILITY OF SUCH DAMAGE.                                            #
+# ----------------------------------------------------------------------------#
+
+import json
+import logging
+import time
+import threading
+import ConfigParser
+import os
+import Queue
+from time import gmtime, strftime
+from threading import Lock
+
+from liota.dccs.dcc import DataCenterComponent, RegistrationFailure, SetPropertiesFailure, CreateRelationshipFailure
+from liota.lib.protocols.helix_protocol import HelixProtocol
+from liota.entities.metrics.metric import Metric
+from liota.lib.utilities.utility import LiotaConfigPath, getUTCmillis, mkdir, read_liota_config
+from liota.lib.utilities.si_unit import parse_unit
+from liota.entities.metrics.registered_metric import RegisteredMetric
+from liota.entities.registered_entity import RegisteredEntity
+
+log = logging.getLogger(__name__)
+
+timeout = 300
+
+class Request:
+    def __init__(self, guid, operation_id, user_queue):
+        self.guid = guid
+        # assume operation_id will be unique in one process
+        self.operation_id = operation_id
+        self.user_queue = user_queue
+
+class Iotcv2(DataCenterComponent):
+    """ The implementation of VMware IoTC V2 Data Center Component
+
+    """
+
+    def __init__(self, con):
+        log.info("Logging into IOTC V2 DCC")
+        self.comms = con
+        if not self.comms.identity.username:
+            log.error("Username not found")
+            raise ValueError("Username not found")
+        elif not self.comms.identity.password:
+            log.error("Password not found")
+            raise ValueError("Password not found")
+        thread = threading.Thread(target=self.comms.receive)
```

Comment:
Why are we starting two thread's here? One on line 79 and another on line 89? We only need one thread for receiving the messages, right?

Reply:
Now we must have 2 threads, which can allow us to support multiple register calls in parallel without losing any responses. Each register call will have a internal queue to block and wait on for responses.
The 2 threads are:
1 is receiving message thread on Line 79 to put received messages into recv_msg_queue; 
2 is to check received messages in recv_msg_queue and dispatch them into each call's response queue.